### PR TITLE
Updates tf interface to be TensorFlow 2.0 compatible

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

* TensorFlow 2.0 was released today, this update ensures PennyLane continues to work with the latest TF version.

**Description of the Change:**

* Renames the tfe interface to `interface="tf"`. This corresponds to TensorFlow 2.0 making eager execution the default user interface.

* TensorFlow 2.0 does not support Python 3.5, so the 3.5 test suite skips the "tf" tests

* TensorFlow has removed all gradient functions with the exception of `tf.GradientTape`

**Benefits:** Support for TF 2.0

**Possible Drawbacks:**

* Note that PennyLane does not work with autograph (`tf.function`)

* Strawberry Fields will currently not import when tensorflow 2.0 is installed, due to the `tf.log` math function being removed.

* TensorFlow 2.0 breaks the `capfd` fixture in pytest.

**Related GitHub Issues:** n/a

### CODE REVIEW

**Comment by josh146** [comment]
Note: we could have separate `tfe` and `tf2` interfaces if we want to support both tf1 and tf2. Thoughts?

File: tests/test_about.py [hunk 1]
```diff
@@ -15,17 +15,25 @@
...
     about: Tests if the about string prints correct.
     """
     qml.about()
-    out, err = capfd.readouterr()
+
+    with patch("sys.stdout", new=StringIO()) as redirect:
+        qml.about()
+
+    out = redirect.getvalue().strip()
+
```
Comment (co9olguy) [inline tests/test_about.py]: What's the story here? 🤔

File: pennylane/interfaces/tf.py [hunk 2]
```diff
@@ -14,28 +14,24 @@
...
+    To use the TensorFlow interface in PennyLane, you must first install TensorFlow.
 
-    >>> import tensorflow as tf
-    >>> import tensorflow.contrib.eager as tfe
-    >>> tf.enable_eager_execution()
+    Note that this interface **only** supports TensorFlow versions >=1.12
+    (including version 2.0) in eager execution mode!
 
-Using the TensorFlow eager execution interface is easy in PennyLane --- let's consider a few ways
+Using the TensorFlow interface is easy in PennyLane --- let's consider a few ways
```
Comment (co9olguy) [inline pennylane/interfaces/tf.py]:
```suggestion
Using the TensorFlow interface is easy in PennyLane---let's consider a few ways
```

File: pennylane/interfaces/tf.py [hunk 3]
```diff
@@ -44,25 +40,25 @@
...
         qml.PhaseShift(theta, wires=0)
         return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
 
-The QNode ``circuit()`` is now a TensorFlow-capable QNode, accepting ``tfe.Variable`` objects
+The QNode ``circuit()`` is now a TensorFlow-capable QNode, accepting ``Variable`` objects
 as input, and returning ``tf.Tensor`` objects.
 
->>> phi = tfe.Variable([0.5, 0.1])
->>> theta = tfe.Variable(0.2)
+>>> phi = Variable([0.5, 0.1])
```
Comment (co9olguy) [inline pennylane/interfaces/tf.py]: same question as above

File: pennylane/decorator.py [hunk 4]
```diff
@@ -144,6 +144,9 @@ def qfunc_decorator(func):
         if interface == 'torch':
             return qnode.to_torch()
 
+        if interface == 'tf':
+            return qnode.to_tf()
+
         if interface == 'tfe':
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/decorator.py]: TensorFlow 1.X must continue to use `'tfe'` (this was the convention used to import eager mode in 1.x from contrib, hence why it was chosen).

In TensorFlow 2.X, eager mode is the default mode when importing TensorFlow as tf.

All code is exactly the same (hence why the `to_tf` function continues to work), the only difference is the import statements required.

File: pennylane/interfaces/__init__.py [hunk 5]
```diff
@@ -50,5 +50,5 @@
 
     interfaces/numpy
     interfaces/torch
-    interfaces/tfe
+    interfaces/tf
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/__init__.py]: This is something that I realised; the code required for the 1.X 'tfe' and 2.0 'tf' interfaces is exactly the same, the only thing that differs is the import statements.

Rather than duplicating codes and tests, we should have 1 doc page, 1 test, and 1 module for the `tf` interface, but make clear to users that they must use `'tfe'` if using tf 1.0.

At some point we can quietly deprecate/remove 'tfe'.

File: pennylane/interfaces/tf.py [hunk 6]
```diff
@@ -14,28 +14,24 @@
...
-
-    This interface **only** supports TensorFlow in eager execution mode! This can be set
-    by running the following commands at the beginning of your PennyLane script/program:
+    To use the TensorFlow interface in PennyLane, you must first install TensorFlow.
 
-    >>> import tensorflow as tf
-    >>> import tensorflow.contrib.eager as tfe
-    >>> tf.enable_eager_execution()
+    Note that this interface **only** supports TensorFlow versions >=1.12
+    (including version 2.0) in eager execution mode!
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]: Yep. The only differences that I can tell:

### TF 1.X:

```python
import tensorflow as tf
import tensorflow.contrib.eager as tfe
tf.executing_eagerly()
Variable = tfe.Variable
```


### TF 2.X:

```python
import tensorflow as tf
from tensorflow import Variable
```

File: pennylane/interfaces/tf.py [hunk 7]
```diff
@@ -44,25 +40,25 @@
...
+    @qml.qnode(dev, interface='tf')
     def circuit(phi, theta):
         qml.RX(phi[0], wires=0)
         qml.RY(phi[1], wires=1)
         qml.CNOT(wires=[0, 1])
         qml.PhaseShift(theta, wires=0)
         return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
 
-The QNode ``circuit()`` is now a TensorFlow-capable QNode, accepting ``tfe.Variable`` objects
+The QNode ``circuit()`` is now a TensorFlow-capable QNode, accepting ``Variable`` objects
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]: I'm unsure the best way to write this so that it is clear. The import paths are different depending on the version of TensorFlow:

* 1.X: `tensorflow.contrib.eager.Variable`
* 2.X: `tensorflow.Variable`

I thought dropping the full path would make it clear, it's up to the user to look up the tf docs to find out where Variable is.

File: tests/test_about.py (hunk 1 above)
Comment (josh146) [reply tests/test_about.py]: TensorFlow uses the abseil package, which replaces the default Python logging interface as v0.7.

There were lots of complaints about this, because it breaks peoples own configured logging modules. Another side effect I discovered was that it breaks the `capfd` and `caplog` fixtures in pytest, which would cause the `test_about.py` and `test_configuration.py` tests to crash with no traceback.

They just pushed a v0.8 of abseil which attempts to fix this (so now `caplog` works), but as far as I can tell `capfd` is still broken by v0.8 :slightly_frowning_face:

File: pennylane/interfaces/tf.py [hunk 8]
```diff
@@ -181,9 +183,9 @@ def circuit(phi, theta):
 The final weights and circuit value are:
 
 >>> phi
-<tf.Variable 'Variable:0' shape=(2,) dtype=float64, numpy=array([ 1.04719755,  0.1       ])>
+<Variable 'Variable:0' shape=(2,) dtype=float64, numpy=array([ 1.04719755,  0.1       ])>
 >>> theta
-<tf.Variable 'Variable:0' shape=() dtype=float64, numpy=0.20000000000000001>
+<Variable 'Variable:0' shape=() dtype=float64, numpy=0.20000000000000001>
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]: Oops, good catch.
```suggestion
<tf.Variable 'Variable:0' shape=() dtype=float64, numpy=0.20000000000000001>
```

File: pennylane/interfaces/tf.py [hunk 9]
```diff
@@ -181,9 +183,9 @@ def circuit(phi, theta):
 The final weights and circuit value are:
 
 >>> phi
-<tf.Variable 'Variable:0' shape=(2,) dtype=float64, numpy=array([ 1.04719755,  0.1       ])>
+<Variable 'Variable:0' shape=(2,) dtype=float64, numpy=array([ 1.04719755,  0.1       ])>
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]:
```suggestion
<tf.Variable 'Variable:0' shape=(2,) dtype=float64, numpy=array([ 1.04719755,  0.1       ])>
```

File: .travis.yml [hunk 10]
```diff
@@ -5,18 +5,22 @@ matrix:
...
+    env: TF=1
 env:
   global:
   - COVERALLS_PARALLEL=true
   - LOGGING=info
 install:
 - pip install -r requirements.txt
-- pip install wheel pytest pytest-cov --upgrade
-- pip install tensorflow==1.13.1
+- pip install wheel pytest pytest-cov absl-py --upgrade
```
Comment (antalszava) [inline .travis.yml]: The test suite seems to run successfully without installing "absl-py". Maybe it would be nice to have a test case where in the example an error would arise without having absl-py installed.

File: tests/test_about.py [hunk 11]
```diff
@@ -15,17 +15,25 @@
...
 import pytest
-import re
 
 
-def test_about(capfd):
+def test_about():
     """
     about: Tests if the about string prints correct.
     """
     qml.about()
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_about.py]:
```suggestion
```

File: pennylane/interfaces/tf.py [hunk 12]
```diff
@@ -112,19 +108,20 @@ def circuit(phi, theta):
...
-    phi = tfe.Variable([0.5, 0.1])
-    theta = tfe.Variable(0.2)
+    phi = Variable([0.5, 0.1])
+    theta = Variable(0.2)
 
-    grad_fn = tfe.implicit_value_and_gradients(circuit)
-    result, [(phi_grad, phi_var), (theta_grad, theta_var)] = grad_fn(phi, theta)
+    with tf.GradientTape() as tape:
+        loss = tf.abs(circuit(phi, theta) - 0.5) ** 2
+        phi_grad, theta_grad = tape.gradient(loss, [phi, theta])
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]: Good catch!

File: .travis.yml (hunk 10 above)
Comment (josh146) [reply .travis.yml]: If you don't have TensorFlow installed, the test suite runs fine (without `absl-py`). However, installing TensorFlow drags in `absl-py` as a dependency, which then breaks pytest.

The latest version of `absl-py` (v0.8) breaks pytest slightly less than v0.7.1, though, which is nice!

The only reason it is added here is that TravisCI has a habit of caching python wheels, and not upgrading packages unless you request it.

File: pennylane/decorator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/decorator.py]: Realised my previous comment was incorrect.

All code is exactly the same (hence why the `to_tf` function continues to work), the only difference is the import statements required. This is caught with a try-except, so the same method works for both versions 1.X and 2.X of TF.

I only kept `'tfe'` for backwards compatibility.

File: pennylane/interfaces/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/__init__.py]: Nevermind, lets just remove 'tfe'.

File: setup.py [hunk 13]
```diff
@@ -29,7 +29,8 @@
 
 extra_requirements = {
     'pytorch':  ["torch", "torchvision"],
-    'tf':  ["tensorflow>=1.12"],
+    'tf':  ["tensorflow>=2.0"],
+    'tfe':  ["tensorflow==1.14"],
```
[the remark this answers is not in the document]
Comment (josh146) [reply setup.py]: Removed.

File: pennylane/interfaces/tf.py (hunk 3 above)
Comment (josh146) [reply pennylane/interfaces/tf.py]: The import paths are different depending on the version of TensorFlow:

* `tensorflow.contrib.eager.Variable`
* `tensorflow.Variable`

My feeling is that explicitly having `Variable` (assuming the user knows where to import it from) helps avoid any confusion

File: pennylane/interfaces/tf.py [hunk 14]
```diff
@@ -14,28 +14,24 @@
...
 
-**Module name:** :mod:`pennylane.interfaces.tfe`
+**Module name:** :mod:`pennylane.interfaces.tf`
 
-.. currentmodule:: pennylane.interfaces.tfe
+.. currentmodule:: pennylane.interfaces.tf
 
 Using the TensorFlow interface
-------------------------------
+----------------------------------
```
Comment (josh146) [inline pennylane/interfaces/tf.py]:
```suggestion
------------------------------
```

File: setup.py [hunk 15]
```diff
@@ -27,11 +27,6 @@
     "semantic_version==2.6"
 ]
 
-extra_requirements = {
-    'pytorch':  ["torch", "torchvision"],
-    'tf':  ["tensorflow>=1.12"],
-}
-
```
Comment (josh146) [inline setup.py]: These were never used in our installation guide, and will now no longer work, as pytorch has moved to hosting their own pip repository.

**Comment by josh146** [comment]
All comments/suggested changes have been applied, and this is now ready for re-review :slightly_smiling_face:

File: pennylane/interfaces/tf.py [hunk 16]
```diff
@@ -112,19 +108,20 @@ def circuit(phi, theta):
...
 
-    phi = tfe.Variable([0.5, 0.1])
-    theta = tfe.Variable(0.2)
+    phi = Variable([0.5, 0.1])
+    theta = Variable(0.2)
 
-    grad_fn = tfe.implicit_value_and_gradients(circuit)
-    result, [(phi_grad, phi_var), (theta_grad, theta_var)] = grad_fn(phi, theta)
+    with tf.GradientTape() as tape:
+        loss = circuit(phi, theta)
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/interfaces/tf.py]:
```suggestion
        # Use the circuit to calculate the loss value
        loss = circuit(phi, theta)
```
